### ISSUE TITLE
Remove non-API calls in C

### ISSUE THREAD

**Issue by MichaelChirico** [Rdatatable/data.table#6180]
We are getting these NOTEs on CRAN, e.g.

Found non-API calls to R: ‘SETLENGTH’, ‘SET_S4_OBJECT’,
      ‘SET_TRUELENGTH’, ‘UNSET_S4_OBJECT’

I'm really skeptical about the SETLENGTH and SET_TRUELENGTH parts, which have long been a key underlying feature of data.table internals. ~I propose we try and keep using those unless CRAN really forces our hands~ CRAN is really forcing our hands :). It be really great if someone wanted to do a performance comparison of the package with/without those calls.

Anyway, it should be easier to remove the non-API S4 calls.

The list does keep changing as this is an actively evolving piece of r-devel. Here's the current list of things we should (eventually) address:

- [x] LEVELS
- [x] OBJECT
- [x] Rf_GetOption
- [x] NAMED
- [x] SET_S4_OBJECT
- [x] SET_TYPEOF
- [x] STRING_PTR
- [x] UNSET_S4_OBJECT
- [x] isFrame
- [x] SETLENGTH
- [x] SET_TRUELENGTH
- [x] SET_GROWABLE_BIT
- [x] TRUELENGTH
- [x] findVar
- [x] SET_ATTRIB
- [x] ATTRIB

**Comment by tdhock** [comment]
there is a web page which lists all of the API calls, grouped into 3 categories
https://yutannihilation.github.io/R-fun-API/
see also the R-devel thread [which is linked from that page](https://stat.ethz.ch/pipermail/r-devel/2024-June/083449.html)

**Comment by ecoRoland2** [comment]
You should probably be proactive and contact R-core regarding your need of an API to `SETLENGTH` and `SET_TRUELENGTH`. I would hate seeing slower performance in data.table. It could cause some difficult decisions at my day job.

**Comment by MichaelChirico** [comment]
if this is important for you (or anyone else reading) please take the time to construct a performance comparison.

**Comment by ecoRoland2** [comment]
I would but unfortunately I think you need at least basic knowledge of C to do that.

**Comment by ben-schwen** [comment]
AFAIU we can replace `SETLENGTH` calls with `Rf_lengthgets` calls and reassignment. But there is no safe way to handle `SET_TRUELENGTH` right?

**Comment by tdhock** [comment]
Where is `SET_TRUELENGTH` used? https://github.com/search?q=repo%3ARdatatable%2Fdata.table%20SET_TRUELENGTH&type=code
one place is in rbindlist which is a pretty important function for performance.
@Anirban166 can you please construct a PR with a single change `SET_TRUELENGTH` -> `Rf_lengthgets` in rbindlist along with an atime performance test?

**Comment by ben-schwen** [comment]
> Where is `SET_TRUELENGTH` used?
We use it as part of memrecycle and subsetting so its basically in most operations :D

**Comment by Anirban166** [comment]
> @Anirban166 can you please construct a PR with a single change `SET_TRUELENGTH` -> `Rf_lengthgets` in rbindlist along with an atime performance test?

Sure, but I'm wondering what the tests will be focused on (since this affects multiple functions) and then what about `SETLENGTH`?

**Comment by MichaelChirico** [comment]
> AFAIU we can replace SETLENGTH calls with Rf_lengthgets calls and reassignment.

I don't see `lengthgets` mentioned in [WRE](https://cran.r-project.org/doc/manuals/R-exts.html#The-R-API).

By the same definition, touching S4 objects from C is also non-API! :thinking:

**Comment by MichaelChirico** [comment]
I see, `lengthgets` is just what `length<-` calls under the hood. But that induces a copy:

```r
a = 1:10
address(a)
# [1] "0x5651b8564b30"
length(a) <- 20
address(a)
# [1] "0x5651b8ab5c68"
```

**Comment by MichaelChirico** [comment]
It looks like more of our usage has been marked as non-API in the meantime:

```
Result: NOTE 
  File ‘data.table/libs/data_table.so’:
    Found non-API calls to R: ‘LEVELS’, ‘NAMED’, ‘SETLENGTH’,
      ‘SET_GROWABLE_BIT’, ‘SET_S4_OBJECT’, ‘SET_TRUELENGTH’,
      ‘UNSET_S4_OBJECT’
```

new: `LEVELS()`, `NAMED()`, `SET_GROWABLE_BIT()`.

**Comment by tdhock** [comment]
there is a new section in WRE https://cran.r-project.org/doc/manuals/r-devel/R-exts.html#Moving-into-C-API-compliance

**Comment by Jean-Romain** [comment]
@tdhock  there is nothing about `SETLENGTH` and `SET_TRUELENGTH` in this new section. I'm in trouble with that one too. I can remove the feature that uses it but I'm wondering how `data.table` team will handle as it is a key feature for data.table.

**Comment by jangorecki** [comment]
I can imagine group by, when not gforce optimized, is setting true length as well. It allocates memory for a biggest group and then reuses that space for smaller groups as well.
Similarly frollapply is doing the same for adaptive=TRUE (possibly not yet in master).

**Comment by ben-schwen** [comment]
This issue also makes our CI fail since we get now a new 3rd NOTE: `Compiled code should not call non-API entry points in R.` Should we alter the CI to allow 3 notes on dev?

**Comment by jangorecki** [comment]
Yes, for the moment till we address this issue. CI is used to spot new issues, this one is already known.

**Comment by MichaelChirico** [comment]
After #6312 and #6313 I'll bump the milestone here to the next release, I think we've put in good effort and definitely applied the fixes they recommend in R-exts. The others remain more difficult to satisfy & can be handled later.

**Comment by jangorecki** [comment]
Afair growable vectors are part of the public api, and those calls which apparently should be removed now are essential tools to work with growable vectors. That may be some argument to request official support for those.

**Comment by jangorecki** [comment]
Uh then it is possibly not a public api then. I had that impression when recalling how well it was announced when got implemented.

**Comment by jangorecki** [comment]
I don't think you mean growable vectors. AFAIU all three functions you mentioned were implemented in DT before growable vector happened in R.

**Comment by SebKrantz** [comment]
Ok, yes I mean the manipulation of this visible and true length with `SETLENGTH` and `SET_TRUELENGTH` - you use that to overallocate data.tables and make them behave like normal data.frames at the R-level which is effectively creating a growable vector - the vector is fixed but the overallocated part is hidden. In `chmatch()` you use the same functions to use the CHARSXP table as a hash table. Anyway, it seems to me `:=` wouldn't be possible in an API conform way without this length manipulation - thus my question about R-core engagement.

**Comment by aitap** [comment]
Over-allocated vectors (including lists since R-4.3.0) can be implemented using ALTREP in something like 60-100 lines of code (separate implementation for [each](https://aitap.codeberg.page/R-api/#R_005fmake_005faltstring_005fclass) primitive [type](https://aitap.codeberg.page/R-api/#R_005fmake_005faltlist_005fclass), unfortunately). I expect it to be hard but not impossible to (1) refactor `data.table`'s use of `SETLENGTH`, `TRUELENGTH`, `SET_GROWABLE_BIT` into an abstract "over-allocated growable vector" interface and (2) reimplement it in ALTREP `#if R_VERSION >= R_Version(4,3,0)`.

Use of `SET_TRUELENGTH` in string matching functions as a free integer-sized field in every `CHARSXP` will be much harder to get rid of. I'm not seeing any other fields in a `CHARSXP` that could be used in API-compliant way for this purpose, which leaves hash tables, which [have different performance characteristics](https://rdatatable.gitlab.io/data.table/reference/chmatch.html#details).

**Comment by SebKrantz** [comment]
It's a nice idea, although it would significantly reduce backwards compatibility by one major version (data.table currently depends R >= 3.3.0) and I'm not sure what consequences ALTREP would have for the interopertability of data.table, particularly for package C code meant to manipulate standard data frames (and for storage as rds and reading if data.table is not loaded)? In any case, if maintainers are unsure how to proceed, perhaps invite R-core to the conversation (e.g. Thomas Kalibera (kalibera) or Luke Tierney (ltierney)). Seems also the vctrs team has not found a solution yet (https://github.com/r-lib/vctrs/issues/1933) so it seems this is not a small issue (and we're talking about probably a third or quarter of CRAN packages that depends on data.table and vctrs, so it must be of concern to R-core that a good solution is found here)

**Comment by aitap** [comment]
Part of the idea is to keep both implementations and use the original one for R &in; [3.3.0; 4.3.0) and the ALTREP one for R &ge; 4.3.0. Maintenance burden, yes, but the old code is already quite stable, and we're not going to have another R-4.2.x version.

The C code, even if it requests raw data pointers, shouldn't notice anything amiss: since the ALTREP class already carries a full-blown over-allocated object of the same type inside it, it can return its (writeable) data pointer when requested.

Serialization will have similar properties to what we have right now: if an ALTREP class does _not_ implement a [`Serialize`](https://aitap.codeberg.page/R-api/#R_005fset_005faltrep_005f_002e_002e_002e_005fmethod) method, R will serialize the contents of the object as if it was an ordinary value, requiring a `setDT()` when it's reloaded. If we really want to, we can implement `Serialize` later, so that the ALTREP objects would come out already over-allocated.

**Comment by TysonStanley** [comment]
Moved to 1.18.0 (but can ship with a patch as well).

**Comment by jangorecki** [comment]
@ben-schwen as those started to be warnings rather than notes, we should aim for prompt fixing those. Do you think you will have enough time to work on those in the next week?

**Comment by ben-schwen** [comment]
@jangorecki I just checked with other packages that still have non API calls such as [vroom](https://cran.r-project.org/web/checks/check_results_vroom.html) and it appears that the warning is triggered by 
> These entry points may be removed soon:

Hence, as soon as we submit patch which removes the `OBJECT` entry point, we should be safe. 

Nevertheless Luke mentioned that they want to turn on the warnings for `TRUELENGTH`, `SET_TRUELENGTH`, etc. in December.

**Comment by TysonStanley** [comment]
Moving the remaining aspects of this to 1.19.0 for now, may have to do a 1.18.2 depending on how fast CRAN changes those NOTES to WARNS
